contributors/design-proposals/controller-ref.md

```diff
@@ -1,101 +1,411 @@
-
-If the owner's selector or owned object's labels change, the owning controller will be responsible for orphaning (clearing `Controller` field in the `OwnerReference` and/or deleting `OwnerReference` altogether) objects, after which adoption procedure (setting `Controller` field in one of `OwnerReferencec` and/or adding new `OwnerReferences`) might occur, if another controller has a selector matching.
-
-For debugging purposes we want to add an `adoptionTime` annotation prefixed with `kubernetes.io/` which will keep the time of last controller ownership transfer.
-
-# Upgrade/downgrade procedure
-
-Because `ControllerRef` will be a part of `OwnerReference` effort it will have the same upgrade/downgrade procedures.
-
-# Orphaning/adoption
-
-Because `ControllerRef` will be a part of `OwnerReference` effort it will have the same orphaning/adoption procedures.
-
-Controllers will orphan objects they own in two cases:
-* Change of label/selector causing selector to stop matching labels (executed by the controller)
-* Deletion of a controller with `Orphaning=true` (executed by the GarbageCollector)
-
-We will need a secondary orphaning mechanism in case of unclean controller deletion:
-* GarbageCollector will remove `ControllerRef` from objects that no longer points to existing controllers
-
-Controller will adopt (set `Controller` field in the `OwnerReference` that points to it) an object whose labels match its selector iff:
-* there are no `OwnerReferences` with `Controller` set to true in `OwnerReferences` array
-* `DeletionTimestamp` is not set
-and
-* Controller is the first controller that will manage to adopt the Pod from all Controllers that have matching label selector and don't have `DeletionTimestamp` set.
-
-By design there are possible races during adoption if multiple controllers can own a given object.
-
-To prevent re-adoption of an object during deletion the `DeletionTimestamp` will be set when deletion is starting. When a controller has a non-nil `DeletionTimestamp` it won't take any actions except updating its `Status` (in particular it won't adopt any objects).
-
-# Implementation plan (sketch):
-
-* Add API field for `Controller`,
-* Extend `OwnerReference` adoption procedure to set a `Controller` field in one of the owners,
-* Update all affected controllers to respect `ControllerRef`.
-
-Necessary related work:
-* `OwnerReferences` are correctly added/deleted,
-* GarbageCollector removes dangling references,
-* Controllers don't take any meaningful actions when `DeletionTimestamps` is set.
+A ControllerRef is thus defined as an OwnerReference with `Controller=true`.
+Each object may have at most one ControllerRef in its list of OwnerReferences.
+The validator for OwnerReferences lists will fail any update that would violate
+this invariant.
```

Comment:
@erictune In case of SparkJob, we're only planning on using ControllerRefs for the form (SparkJob TPR -> driver-pod -> executor pods). All other resources created, i.e. secrets and services will have OwnerReferences pointing back to the TPR.